src/hotspot/share/gc/shenandoah/shenandoahFreeSet.hpp

```diff
@@ -27,70 +28,364 @@
+class ShenandoahRegionPartitions {
+
+private:
+  const ssize_t _max;           // The maximum number of heap regions
+  const size_t _region_size_bytes;
+  const ShenandoahFreeSet* _free_set;
+  // For each partition, we maintain a bitmap of which regions are affiliated with his partition.
+  ShenandoahSimpleBitMap _membership[UIntNumPartitions];
+
+  // For each partition, we track an interval outside of which a region affiliated with that partition is guaranteed
+  // not to be found. This makes searches for free space more efficient.  For each partition p, _leftmosts[p]
+  // represents its least index, and its _rightmosts[p] its greatest index. Empty intervals are indicated by the
+  // canonical [_max, -1].
+  ssize_t _leftmosts[UIntNumPartitions];
+  ssize_t _rightmosts[UIntNumPartitions];
+
+  // Allocation for humongous objects needs to find regions that are entirely empty.  For each partion p, _leftmosts_empty[p]
+  // represents the first region belonging to this partition that is completely empty and _rightmosts_empty[p] represents the
+  // last region that is completely empty.  If there is no completely empty region in this partition, this is represented
+  // by the canonical [_max, -1].
+  ssize_t _leftmosts_empty[UIntNumPartitions];
+  ssize_t _rightmosts_empty[UIntNumPartitions];
+
+  // For each partition p, _capacity[p] represents the total amount of memory within the partition at the time
+  // of the most recent rebuild, _used[p] represents the total amount of memory that has been allocated within this
+  // partition (either already allocated as of the rebuild, or allocated since the rebuild).  _capacity[p] and _used[p]
+  // are denoted in bytes.  Note that some regions that had been assigned to a particular partition at rebuild time
+  // may have been retired following the rebuild.  The tallies for these regions are still reflected in _capacity[p]
+  // and _used[p], even though the region may have been removed from the free set.
+  size_t _capacity[UIntNumPartitions];
+  size_t _used[UIntNumPartitions];
+  size_t _region_counts[UIntNumPartitions];
+
+  // Shrink the intervals associated with partition when region idx is removed from this free set
+  inline void shrink_interval_if_boundary_modified(ShenandoahFreeSetPartitionId partition, ssize_t idx);
+
+  // Shrink the intervals associated with partition when regions low_idx through high_idx inclusive are removed from this free set
+  inline void shrink_interval_if_range_modifies_either_boundary(ShenandoahFreeSetPartitionId partition,
+                                                                ssize_t low_idx, ssize_t high_idx);
+  inline void expand_interval_if_boundary_modified(ShenandoahFreeSetPartitionId partition, ssize_t idx, size_t capacity);
+
+  void dump_bitmap_row(ssize_t idx) const;
+  void dump_bitmap_range(ssize_t start_idx, ssize_t end_idx) const;
+  void dump_bitmap_all() const;
```

Comment:
In the first case you can may be say `region_idx` instead of `idx` ?

Also, it seems `dump_bitmap()` would be sufficient (at least for the case of `dump_bitmap_all()`, although in general for all).

However, given that these are private methods that will only be found in the implementation and nowhere else in its clients' code, may be current naming is ok (indeed, it looks like the first two are work functions in service of the last).

PS: I don't see any current usage of `dump_bitmap_all()`. May be you wanted it in some debug tracing code or something, or keeping it for possible future debugging usage? (Or may be my naive web-page search of this review page is naive/faulty.) It's perhaps useful to keep around for debug/tracing though.